### ISSUE TITLE
Pinning coreos-assembler in FCOS releases

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#1068]
Right now we have a testing stream and we have a stable stream. We let things bake in testing a few weeks before promoting to stable, except for all of the changes that have gone into COSA, which could be significant. Maybe we should add a tag to cosa on the day a testing release is run and use that same cosa image that was used for testing when we do the stable pipeline run.

**Comment by dustymabe** [comment]
> Another thing we could do is make the cosa tag to use live in https://github.com/coreos/fedora-coreos-config and have `bump-lockfile` also bump it.

I kind of like it.. though I think that means we'd need to create a tag (git tag) pretty much for every commit against the COSA repo and then do builds against that. Do I understand correctly?

> 
> Some advantages of this is that (1) cosa is "locked" by default which improves reproducibility, and (2) we transparently get cosa tag propagation as part of the regular promotion process. A disadvantage is that it introduces more delay between cosa changes and it affecting FCOS builds (though for hotfixes, we always have the override approach at job run time).

**Comment by cgwalters** [comment]
I agree we've hit problems but so far they haven't been too hard to work around.  My instinct here is to keep this very simple to start.  A very simple thing would be (pretty sure I proposed this elsewhere):

- When a FCOS branch succeeds, it updates a tagged cosa build for that branch to point to the cosa used successfully

Then, in a case like https://github.com/coreos/coreos-assembler/pull/2643 we can just temporarily override the stable branch to use that cosa.

**Comment by bgilbert** [comment]
I agree the workarounds haven't been difficult so far.  The problem is that we discover the issues while performing an out-of-cycle release, which is already a time-sensitive and chaotic process.  And there's still the possibility of changes that don't cause build breakage (and maybe don't even cause CI breakage) but ship unbaked functionality directly to stable.

So, I don't have opinions about the technical means used to accomplish it, but I do think we should find a systematic solution here.

**Comment by dustymabe** [comment]
Note: Most likely a solution to this will require us to have our multi-arch COSA build/pushing (https://github.com/coreos/fedora-coreos-tracker/issues/1027) fleshed out and working.

**Comment by jlebon** [comment]
> > Another thing we could do is make the cosa tag to use live in [coreos/fedora-coreos-config](https://github.com/coreos/fedora-coreos-config) and have `bump-lockfile` also bump it.
> 
> I kind of like it.. though I think that means we'd need to create a tag (git tag) pretty much for every commit against the COSA repo and then do builds against that. Do I understand correctly?

Yeah, I think so. We'll have to investigate the [Quay.io auto-GC options](https://docs.projectquay.io/use_quay.html#tag-expiration). Or maybe as you say if we do our own building and pushing, it wouldn't be hard to have that process manage the tags for us.

**Comment by cgwalters** [comment]
I just did
```
$ oc tag coreos-assembler:latest coreos-assembler:stable                                                                                                                                                                                        04/28/2022 04:54:38 PM
Tag coreos-assembler:stable set to coreos-assembler@sha256:a7441ebcd415b42c871e8f52c81bc1a194826f7e7f13ba506aa13e397b4541fa.
$
```

What I'm thinking actually is that we should have a controller which queries for which coreos-assembler was used to build the FCOS stable job, and updates that tag or so.

**Comment by dustymabe** [comment]
> Note: Most likely a solution to this will require us to have our multi-arch COSA build/pushing (#1027) fleshed out and working.

Noting that at least this part should be unblocked now since https://github.com/coreos/fedora-coreos-tracker/issues/1027 is completed.

**Comment by jlebon** [comment]
Picking this up again since it's been unblocked now. @cgwalters' [strawman](https://github.com/coreos/fedora-coreos-tracker/issues/1068#issuecomment-1014818291) seems like it would be simple enough to start. Translating it into more concrete terms:
1. In the FCOS pipeline `release` job, we push a Quay.io tag like `quay.io/coreos-assembler/coreos-assembler:fcos-${params.STREAM}`.
2. In the FCOS pipeline `build` job, we look up the correct tag to use for the stream being built (this would re-implement [this knowledge](https://github.com/coreos/fedora-coreos-streams/blob/5790a4f4c790353959450417f2561ec576d15393/.github/workflows/promote-config.yml#L24-L33), which ideally we'd export... into e.g. a JSON file maybe in the pipeline repo?).

The advantage of this over the cosa branches approach is that we're re-using the exact same image, whereas rebuilding a cosa git branch currently can yield different package sets each time. (Really tempting to start locking cosa...) The disadvantage is that if we really need to backport a patch, it wouldn't be as streamlined an operation. But I think we've very rarely had to do that; the primary issues have just been from using `:main` instead of an older image.

**Comment by dustymabe** [comment]
Part of the problem is that the cosa tag in quay could get updated in the middle of a run, so pushing a quay.io tag in the `release` job would be too late I think.

How about we create a new toplevel job that then triggers `build.Jenkinsfile`. This new toplevel job is only run manually (i.e. when we manually run prod builds). This job will:

- For the `testing` run (same for `next`):
    - create a `quay.io/coreos-assembler/coreos-assembler-staging:fcos-testing` tag based on current `quay.io/coreos-assembler/coreos-assembler:main`.
    - launch the `build` job for `testing` using this new container tag
- For the `stable` run:
    - create a `quay.io/coreos-assembler/coreos-assembler-staging:fcos-stable` tag based on current `quay.io/coreos-assembler-staging/coreos-assembler:fcos-testing`.


This kind of requires that `stable` run before `testing` in the triple release workflow, otherwise the `fcos-testing` tag would have been updated with today's update rather than the one from two weeks ago. One idea here is to run all 3 builds as part of this job (i.e. one job triggers the build job for `stable`, `testing` and `next`). That way the ordering is enforced.

We could also verify that the prod streams use the right COSA image by codifying it in the `build.Jeninsfile` job.

There are still some details to be worked out here, but I think their is a working solution in here somwehere.

**Comment by bgilbert** [comment]
I'm not thrilled by the ordering dependency in that proposal, and we'd also have to handle out-of-cycle releases, which shouldn't update the cosa pins.

Semantically, we're trying to augment the fedora-coreos-config promotion process to pin cosa containers, right?  Could we just do this directly in the f-c-c promotion job for `next` and `testing`?  It could tag cosa `main` to prevent GC, and write the tag into a file in the branch.  Further promotions and out-of-cycle builds would then automatically DTRT, and the development streams could continue to build with `main`.

**Comment by dustymabe** [comment]
> I'm not thrilled by the ordering dependency in that proposal, and we'd also have to handle out-of-cycle releases, which shouldn't update the cosa pins.

My proposal on how to handle that is to update the build job to default to `quay.io/coreos-assembler-staging/coreos-assembler:fcos-${STREAM}` for prod streams.

For out-of-cycle releases you just run the build job directly with the stream selected and all other defaults (basically what you do today).

> 
> Semantically, we're trying to augment the fedora-coreos-config promotion process to pin cosa containers, right? Could we just do this directly in the f-c-c promotion job for `next` and `testing`? It could tag cosa `main` to prevent GC, and write the tag into a file in the branch. Further promotions and out-of-cycle builds would then automatically DTRT, and the development streams could continue to build with `main`.

The reason I think doing it in a pipeline job is best is just because we already have the credentials there and the tools there to do everything. Rather than wiring that up in a GH action.

**Comment by bgilbert** [comment]
> The reason I think doing it in a pipeline job is best is just because we already have the credentials there and the tools there to do everything. Rather than wiring that up in a GH action.

Hmm, I'm not sure I understand that.  I think it's worth a moderate amount of code complexity in exchange for the operational flexibility of running builds in any order.  And the build job proposal sounds like a larger change in any event.  I think my proposal basically reduces to:

```bash
if [[ $src_branch =~ ^testing-devel|next-devel$ ]]; then
    repo="quay.io/coreos-assembler/coreos-assembler"
    manifest=$(skopeo inspect "docker://${repo}:latest" | jq -r .Digest)
    tag=fcos-$(echo "${manifest}" | cut -f2 -d: | cut -c-10)
    skopeo copy -a "docker://${repo}@${manifest}" "docker://${repo}:${tag}"
    echo "${repo}:${tag}" > .cosa-ref
    git add .cosa-ref
fi
```

plus credential handling similar to [this](https://github.com/coreos/actions-lib/blob/a9d25d3b3ddb979145bee155c966f68d4e1b7a80/build-container/action.yml#L98-L110).
